=per #18751 Persistence: Return completed current events stream for cleaned journal

eshepelyuk opened the pull request:
Fix for #18751

patriknw committed:
=rem #18339 Use explicit handshake timeout

* instead of using transport failure detector
* add a new config property akka.remote.handshake-timeout, but
  for netty.tcp and netty.ssl the existing netty.tcp.connection-timeout
  setting will be used
* add test of the timeouts
* mima filter for internal ProtocolStateActor

patriknw committed:
=doc #18636 Correction of mailbox documentation

eshepelyuk committed:
Return empty event stream for completely cleaned journal

eshepelyuk committed:
Return empty event stream for completely cleaned journal

eshepelyuk committed:
Return empty event stream for completely cleaned journal

typesafehub-validator commented:
Hi @eshepelyuk,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Typesafe Contributors License Agreement:

[http://www.typesafe.com/contribute/cla](http://www.typesafe.com/contribute/cla)

jypma committed:
=per #18708 only try further constructors if constructor can't be found

* Any exception in a read journal's constructor can now be
propagated to the caller correctly without being shadowed.

patriknw commented:
OK TO TEST

michaelomann committed:
=doc fix dispatchers docs nr typo, see #18712

markus1189 committed:
=doc Fix sentence in `Actors/Send messages/Ask`

The sentence

  > Gives you a way to avoid blocking.

does not make sense alone, I added it to the previous one.

markus1189 committed:
=doc Missing `by` in Actors/Initialization patterns

markus1189 committed:
=doc Replace "in runtime" with "at runtime"

In the "Fault Tolerance" docs, replace "in runtime" with "at runtime".

eshepelyuk committed:
+per #18559 TCK tests for highestSequenceNr not reset after journal message deletion

markus1189 committed:
=doc Add missing `by` in Routing docs

Add a missing `by` to the Routing docs, Specially Handled Messages section.

Commit:
+doc, per adding clarification about highestSeqenceNr bahaviour after message deletion

Commit:
+doc, per adding clarification about highestSeqenceNr behaviour after message deletion

Commit:
+doc, per adding clarification about highestSeqenceNr behaviour after message deletion

ktoso committed:
=doc explain parallelism-max in more detail

Commit:
=doc =per fixed a few typos from 'Unknow' to 'Unknown'

Commit:
+doc, per adding clarification about highestSequenceNr to
- AsyncRecovery#asyncReadHighestSequenceNr
- AsyncWriteJournal#asyncDeleteMessagesTo

patriknw commented on akka-persistence-query/src/main/scala/akka/persistence/query/journal/leveldb/EventsByPersistenceIdPublisher.scala:
```diff
@@ -181,7 +181,7 @@ private[akka] class CurrentEventsByPersistenceIdPublisher(
     if (highestSeqNr < toSequenceNr)
       toSeqNr = highestSeqNr
-    if (highestSeqNr == 0L || (buf.isEmpty && currSeqNo > toSequenceNr))
+    if (highestSeqNr == 0L || (buf.isEmpty && currSeqNo > toSequenceNr) || currSeqNo == fromSequenceNr)
```
so that extra condition means that nothing was replayed?
do we need the `highestSeqNr == 0L` ?

patriknw replied:
Note that there is a `maxBufSize` which means that the buf might be half filled from a first replay request and then when this is called for the second replay request we must still drain the buffer before completing the stream. Shouldn't the condition be `buf.isEmpty && (currSeqNo > toSequenceNr || currSeqNo == fromSequenceNr)`?

eshepelyuk replied:
I've tried to remove `highestSeqNr == 0L` and use `buf.isEmpty && (currSeqNo > toSequenceNr || currSeqNo == fromSequenceNr)`.
Tests are working, I've added some more cases.

patriknw commented:
do we have the same issue in `CurrentEventsByTagPublisher`?

eshepelyuk commented:
Let me try to add a test for `CurrentEventsByTagPublisher` and check the behavior. If there's the same error - I will try to fix it.
Sounds ok ?

eshepelyuk committed:
+per #18751 Persistence: Return completed current events stream for cleaned journal

eshepelyuk commented:
- modified condition of stream completion for `currentEventsByPersistenceId` query
- test for `currentEventsByTag` after journal cleanup

patriknw commented on akka-persistence-query/src/main/scala/akka/persistence/query/journal/leveldb/EventsByPersistenceIdPublisher.scala:
```diff
@@ -181,7 +181,7 @@ private[akka] class CurrentEventsByPersistenceIdPublisher(
     if (highestSeqNr < toSequenceNr)
       toSeqNr = highestSeqNr
-    if (highestSeqNr == 0L || (buf.isEmpty && currSeqNo > toSequenceNr))
+    if (buf.isEmpty && (currSeqNo > toSequenceNr) || currSeqNo == fromSequenceNr)
```
I'd find it more readable as:

    if (buf.isEmpty && (currSeqNo > toSequenceNr || currSeqNo == fromSequenceNr))

eshepelyuk replied:
Sorry, that was a weird typo. Fixed

patriknw commented on akka-persistence-query/src/test/scala/akka/persistence/query/journal/leveldb/EventsByTagSpec.scala:
```diff
@@ -121,6 +123,23 @@ class EventsByTagSpec extends AkkaSpec(EventsByTagSpec.config)
+      expectMsg(s"${Long.MaxValue}-deleted")
+
+      val src = queries.currentEventsByTag("green", Long.MaxValue)
+      src.map(_.event).runWith(TestSink.probe[Any]).request(1).expectComplete()
```
note that `Long.MaxValue` is the starting offset, i.e. corresponding to `fromSequenceNr` so this test doesn't really do what the title says.

Tags work differently because events are not deleted from the tag-stream when you do `deleteMessages`. There is currently no support for deleting tagged events.

Anyway, I suggest that you change the corresponding condition in the same way in `CurrentEventsByTagPublisher`, given that it does not break any other test.

eshepelyuk commented:
Pushed suggested improvement, deleted irrelevent test.

patriknw commented:
LGTM, thanks. Please create an issue so that we can track this fix. Squash to one commit.

eshepelyuk commented:
I will close this PR and reopen new one.
